### PR TITLE
Caret position on WEB

### PR DESCRIPTION
**Pull request by perunt**
### Fixed Issues

$ https://github.com/Expensify/App/issues/16078
PROPOSAL: GH_LINK_ISSUE(COMMENT)

This PR adds caret tracking for web and desktop platforms, which is an essential step in bringing inline suggestions to the text input

- [x] Verify that no errors appear in the JS console

In this pull request, the methods of the TextInput were expanded, which should not affect any existing functionality. However, to test this pull request, you can use the text inputs anywhere.

For more detailed testing, you can check the value of the onSelectionChange event(in a chat screen). Check the new positionX and positionY properties, which show the coordinates of the caret.
### QA Steps

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the Waiting for Copy label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://user-images.githubusercontent.com/48593211/232459614-b8a1017c-058a-41ac-a036-88a5659eea1b.mp4

</details>

<details>
<summary>Mobile Web - Chrome</summary>

\

https://user-images.githubusercontent.com/48593211/232459126-576d6df7-9b40-496d-b236-0c8b485fedbf.mp4

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://user-images.githubusercontent.com/48593211/232459069-245e41cb-c857-4cd1-a76c-79b9a1446ecb.mp4

</details>

<details>
<summary>Desktop</summary>

</details>

<details>
<summary>iOS</summary>

</details>

<details>
<summary>Android</summary>

</details>

### CODE REVIEW

**Comment by hayata-suenaga** [comment]
@puneetlath since you have the context, if you have time, could you also review this PR?

File: src/components/Composer/index.js [hunk 1]
```diff
@@ -358,28 +398,52 @@ class Composer extends React.Component {
...
         // We're disabling autoCorrect for iOS Safari until Safari fixes this issue. See https://github.com/Expensify/App/issues/8592
         return (
-            <RNTextInput
-                autoComplete="off"
-                autoCorrect={!Browser.isMobileSafari()}
-                placeholderTextColor={themeColors.placeholderText}
-                ref={el => this.textInput = el}
-                selection={this.state.selection}
-                onChange={this.shouldCallUpdateNumberOfLines}
-                onSelectionChange={this.onSelectionChange}
```
Comment (hayata-suenaga) [inline src/components/Composer/index.js:370]: this is a question about old code, but they were we using `this.onSelectionChange`? This supposed to be `this.props.onSelectionChange` right?

File: src/components/Composer/index.js [hunk 2]
```diff
@@ -358,28 +398,52 @@ class Composer extends React.Component {
...
+        const calculationCaretPositionElement = (
+            <View
+                style={{
+                    position: 'absolute',
+                    bottom: -2000,
+                    zIndex: -1,
+                    opacity: 0,
+                }}
+            >
+                <Text
```
Comment (hayata-suenaga) [inline src/components/Composer/index.js:412]: Is this supposed to be `RNTextInput`?

File: src/components/Composer/index.js [hunk 3]
```diff
@@ -190,6 +199,36 @@ class Composer extends React.Component {
...
+        if (this.props.shouldCalculateCaretPosition) {
+            this.setState(
+                {
+                    valueBeforeCaret: event.target.value.slice(
+                        0,
+                        event.nativeEvent.selection.start,
+                    ),
+                },
+
+                () => {
```
[the remark this answers is not in the document]
Comment (mczernek) [reply src/components/Composer/index.js:212]: I'd say it makes sense - in `componentDidUpdate` we'd need to check what really changed. Here we know.

**Review by mczernek — DISMISSED** [review]

File: src/components/Composer/index.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (perunt) [reply src/components/Composer/index.js:212]: Yes, utilizing `onSelectionChangeHandle` helps maintain clarity by allowing you to easily understand what's happening. This way, you don't need to constantly verify if it's truly firing `onSelectionChange` or double-check the data associated with it.

File: src/components/Composer/index.js [hunk 4]
```diff
@@ -358,28 +398,52 @@ class Composer extends React.Component {
...
+                        // so we can get the correct scroll height while calculating the number of lines.
+                        this.state.numberOfLines < this.props.maxLines ? styles.overflowHidden : {},
+                    ]}
                 /* eslint-disable-next-line react/jsx-props-no-spreading */
-                {...propsWithoutStyles}
-                numberOfLines={this.state.numberOfLines}
-                disabled={this.props.isDisabled}
-            />
+                    {...propsWithoutStyles}
+                    onSelectionChange={this.onSelectionChangeHandle}
```
[the remark this answers is not in the document]
Comment (perunt) [reply src/components/Composer/index.js:441]: good catch, i'll change it to `addCursorPositionToSelectionChange `

File: src/components/Composer/index.js (hunk 1 above)
Comment (perunt) [reply src/components/Composer/index.js:370]: I understand your surprise, and you are correct. It should indeed be `this.props.onSelectionChange`

File: src/components/Composer/index.js (hunk 2 above)
Comment (perunt) [reply src/components/Composer/index.js:412]: No, you don't need to use the TextInput component. Instead, the main objective is to place an empty Text component immediately after the primary text. Using TextInput wouldn't allow it to track the final character in the main text input accurately.

**Review by puneetlath — DISMISSED** [review]
Looks good to me.

File: src/components/Composer/index.js (hunk 2 above)
Comment (hayata-suenaga) [reply src/components/Composer/index.js:412]: oh I meant `RNText` but i think it doesn't matter which one to use.

**Review by hayata-suenaga — DISMISSED** [review]
LGTM!

**Comment by perunt** [comment]
@0xmiroslav thanks for letting me know about the missed and new checkbox! Right now, it only works with the main input, but it can be easily transferred to the edit input soon.

**Comment by puneetlath** [comment]
>Right now, it only works with the main input, but it can be easily transferred to the edit input soon.

@perunt so is your plan to do that in a separate PR?

**Comment by perunt** [comment]
Splitting this PR and merging the one in question would help me save time, enabling me to work on the Emoji Suggestion task simultaneously. So yeah, if you don't mind, I would love to split it.

**Comment by puneetlath** [comment]
Ok sounds good. @0xmiroslav will you be able to complete the checklist today?

**Review by 0xmiroslav — COMMENTED** [review]
From my testing, `positionY` has 1px offset between first character and others in the same line.

https://user-images.githubusercontent.com/97473779/235749725-8cf5d671-5b3a-4c56-b4b5-75eeeabab301.mov

File: src/components/Composer/index.js [hunk 5]
```diff
@@ -358,28 +398,52 @@ class Composer extends React.Component {
         propStyles.outline = 'none';
         const propsWithoutStyles = _.omit(this.props, 'style');
 
+        // This code creates a hidden text component that helps track the caret position in the visible input.
+
+        const calculationCaretPositionElement = (
```
[the remark this answers is not in the document]
Comment (perunt) [reply src/components/Composer/index.js]: 👍🏻

**Comment by perunt** [comment]
@0xmiroslav done

**Comment by hayata-suenaga** [comment]
thank you for fixing the merge conflict

> Btw, I noticed another issue with really long strings (like 30 characters or more, or a word that has the length the same as the width of the TextInput). It's a rare case, but it happens because Text and TextInput components have different ways of wrapping text into new lines.

@perunt could you expand on this more? Is my understanding correct that `Text` wrapped inside `TextInput` has a different behavior than the `TextInput`, and that's why there are issues on certain situations?

can you also tell me what are exact issues you observed with long strings? maybe screenshots might be helpful?

**Comment by perunt** [comment]
@hayata-suenaga  let's take a look at some screen recordings. Please note that the red Text component and the composer (TextInput) have the same style.
To detect the cursor position, I created two text components. The first one copies the text before the caret (the red text), while the second one imitates the caret (let's call it the caretText component) and I detect its position.

https://user-images.githubusercontent.com/48593211/236418552-f65b2fb2-c57c-4568-af92-e9eabc745616.mp4


In the first video, you can see that the caretText has no width (in the next video it would be a green area). The text input wraps the line strategy based on the length of the word, which doesn't work for us because TextInput knows that words could be longer, i.e. part of word + caret + another part of word.

https://user-images.githubusercontent.com/48593211/236418702-202a6780-9416-4538-90dd-62fbb96289bb.mp4


The second video demonstrates how it works now. I copied the full word to the Text component to have the same line wrapping strategy. However, with this approach, we have a problem with long words that are wider than the width of the Text input.

**Comment by perunt** [comment]
However, as we have Text (text before caret) + Text (caretText component) in our case and just TextInput as a source of tracking, this component has a different line-wrapping strategy. In the first component(Text+Text) the caretComponent always wraps to the second line if we speak about extremely long words

File: src/components/Composer/index.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (perunt) [reply src/components/Composer/index.js]: 👍🏻

**Comment by hayata-suenaga** [comment]
@perunt thank you for your detailed explanation. I'll wait other's reviews 🙇

**Review by hayata-suenaga — DISMISSED** [review]
don't have detailed review but if the changes are from merging and conforming to prettier then PR LGTM

i'd wait for other's previews though

File: src/components/Composer/index.js [hunk 6]
```diff
@@ -192,6 +201,54 @@ class Composer extends React.Component {
...
+                },
+
+                () => {
+                    const customEvent = {
+                        nativeEvent: {
+                            selection: {
+                                start: event.nativeEvent.selection.start,
+                                end: event.nativeEvent.selection.end,
+
+                                // 4 is a width of the space which we added manually
```
Comment (puneetlath) [inline src/components/Composer/index.js:237]: Should we add this as a constant?

**Review by puneetlath — CHANGES_REQUESTED** [review]
Just some minor comments from me.

File: src/components/Composer/index.js (hunk 6 above)
Comment (perunt) [reply src/components/Composer/index.js:237]: not sure if itwould be used anywhere else but i added it, thanks

File: src/components/Composer/index.js [hunk 7]
```diff
@@ -192,6 +201,54 @@ class Composer extends React.Component {
...
+
+        if (spaceIndex === -1) {
+            return substr;
+        }
+
+        // If there is a space or new line, return the substring up to the space or new line
+        return substr.substring(0, spaceIndex);
+    }
+
+    addCursorPositionToSelectionChange(event) {
```
[the remark this answers is not in the document]
Comment (perunt) [reply src/components/Composer/index.js:220]: 👍🏻

File: src/components/Composer/index.js [hunk 8]
```diff
@@ -192,6 +201,54 @@ class Composer extends React.Component {
         this.textInput.removeEventListener('wheel', this.handleWheel);
     }
 
+    // get characters from the cursor to the next space or new line
```
[the remark this answers is not in the document]
Comment (perunt) [reply src/components/Composer/index.js:204]: 👍🏻

File: src/components/Composer/index.js [hunk 9]
```diff
@@ -1,5 +1,6 @@
+/* eslint-disable rulesdir/onyx-props-must-have-default */
```
[the remark this answers is not in the document]
Comment (perunt) [reply src/components/Composer/index.js]: Initially set a default value, but ran into linter issues. After rechecking, the specific issue is resolved. However, we're facing additional errors with the linter, unrelated to this PR. It seems to conflict with our new Prettier configuration.

**Review by puneetlath — DISMISSED** [review]
Looks good to me. @0xmiroslav over to you.

**Comment by puneetlath** [comment]
@perunt looks like you need to merge main. 

@0xmiroslav think you'll be able to review soon?

**Comment by perunt** [comment]
@0xmiroslav done

**Comment by puneetlath** [comment]
@0xmiroslav please review today if you can!

**Comment by puneetlath** [comment]
>I'm not sure if we face the same problem in the native environment. If we do, it would make sense to create a separate PR and synchronize it across all platforms. If not, then it would be better to include it in this PR.

Sounds like we need to figure out whether it's also a problem on native devices in order to decide.

**Comment by puneetlath** [comment]
@perunt is that a question that you're going to investigate?

**Comment by perunt** [comment]
@puneetlath absolutely! I'll definitely check it out and make sure to mention you. I apologize for the delay as I've had to prioritize a more urgent task. I'll make sure to get to it sometime during the day.

**Comment by puneetlath** [comment]
No worries! Just making sure.

**Comment by hayata-suenaga** [comment]
@perunt did you have a chance to investigate the problem?

**Comment by perunt** [comment]
i just checked it , and it works the same on Android and IOS

https://github.com/Expensify/App/assets/48593211/9701cddc-e278-4f4d-b1ca-ce8467f22064



https://github.com/Expensify/App/assets/48593211/82ca7114-6955-467e-a2ea-6fd15558aab2

**Comment by perunt** [comment]
@hayata-suenaga @puneetlath what do you think, can we handle this measuring and connect all platforms in the separate PR?

**Comment by puneetlath** [comment]
That works for me. @hayata-suenaga @0xmiroslav good with you?

**Comment by puneetlath** [comment]
Ok cool. @0xmiroslav The reviewer checklist check is failing for some reason.

**Review by 0xmiroslav — APPROVED** [review]

**Review by puneetlath — APPROVED** [review]

**Review by hayata-suenaga — APPROVED** [review]
🚀

File: src/components/Composer/index.js [hunk 10]
```diff
@@ -369,29 +430,57 @@ class Composer extends React.Component {
         propStyles.outline = 'none';
         const propsWithoutStyles = _.omit(this.props, 'style');
 
+        // This code creates a hidden text component that helps track the caret position in the visible input.
+        const renderElementForCaretPosition = (
+            <View
+                style={{
+                    position: 'absolute',
+                    bottom: -2000,
```
Comment (narefyev91) [inline src/components/Composer/index.js:438]: this changes made a regression - https://github.com/Expensify/App/issues/26535

**Pull request merged**